src/StreamUtils.bsv

```diff
@@ -0,0 +1,367 @@
+endinterface
+
+interface StreamSplit;
+    interface FifoIn#(DataStream)  inputStreamFifoIn;
+    interface FifoIn#(StreamSize)  splitLocationFifoIn;
+    interface FifoOut#(DataStream) outputStreamFifoOut;
+endinterface
+
+function DataBytePtr convertByteEn2BytePtr (ByteEn byteEn);
+    ByteEn byteEnTemp = byteEn;
+    DataBytePtr ptr = 0;
+    while (byteEnTemp > 0) begin
+        byteEnTemp = byteEnTemp >> 1;
+        ptr = ptr + 1;
+    end
+    return ptr;
+endfunction
+
+function DataStream getEmptyStream ();
+    return DataStream{
+        data: 0,
+        byteEn: 0,
+        isFirst: False,
+        isLast: True
+    };
+endfunction
+
+function DataBitPtr getMaxBitPtr ();
+    return fromInteger(valueOf(DATA_WIDTH));
+endfunction
+
+function DataBytePtr getMaxBytePtr ();
+    return fromInteger(valueOf(BYTE_EN_WIDTH));
+endfunction
+
+// Concat two DataStream frames into one. StreamA.isLast must be True, otherwise the function will return a empty frame to end the stream.
+function Tuple3#(DataStream, DataStream, DataBytePtr) getConcatStream (DataStream streamA, DataStream streamB, DataBytePtr bytePtrA, DataBytePtr bytePtrB);
+    Bool isCallLegally = (streamA.isLast && bytePtrA <= getMaxBytePtr() && bytePtrA > 0);
+    DataBitPtr bitPtrA = zeroExtend(bytePtrA) << fromInteger(valueOf(BYTE_WIDTH_WIDTH));
+
+    // Fill the low PtrA bytes by streamA data
+    Data concatDataA = streamA.data;
+    ByteEn concatByteEnA = streamA.byteEn;
+
+    // Fill the high bytes by streamB data
+    Data concatDataB = streamB.data << bitPtrA;
+    ByteEn concatByteEnB = streamB.byteEn << bytePtrA;
+    Data concatData = concatDataA | concatDataB;
+    ByteEn concatByteEn = concatByteEnA | concatByteEnB;
+
+    // Get the remain bytes of streamB data
+    DataBitPtr resBitPtr = getMaxBitPtr() - bitPtrA;
+    DataBytePtr resBytePtr = getMaxBytePtr() - bytePtrA;
+    Data remainData = streamB.data >> resBitPtr;
+    ByteEn remainByteEn = streamB.byteEn >> resBytePtr;
+
+    // Get if the concat frame is the last
+    Bool isConcatStreamLast = streamB.isLast;
+    DataBytePtr remainBytePtr = 0;
+    if (resBytePtr < bytePtrB ) begin    
+        isConcatStreamLast = False;
+        remainBytePtr = bytePtrB - resBytePtr;
+    end
+     DataStream concatStream = getEmptyStream;
+     DataStream remainStream = getEmptyStream;
+
+    // package the return concatStream and remainStream
+    if(isCallLegally) begin
+        concatStream = DataStream{
+            data: concatData,
+            byteEn: concatByteEn,
+            isFirst: streamA.isFirst,
+            isLast: isConcatStreamLast
+        };
+        remainStream = DataStream{
+            data: remainData,
+            byteEn: remainByteEn,
+            isFirst: False,
+            isLast: True
+        };
+    end
+    return tuple3(concatStream, remainStream, remainBytePtr);
+endfunction
+
+function Action showDataStream (DataStream stream);
```

Comment:
You should make DataStreamderiving FShow, and you can simply display it anyware.

And if you insist on using `showDataStream` , then you may use it many place in your codebase. Why put it in this file?

Reply:
I did not noticed Fshow before. I will replace `showDataStream` funciton with it.